Mathlib/AlgebraicGeometry/Morphisms/AffineAnd.lean

```diff
@@ -0,0 +1,184 @@
+      IsAffineOpen (f ⁻¹ᵁ U) ∧ Q (f.app U) := by
+  simp only [targetAffineLocally, affineAnd_apply, morphismRestrict_app, hQi.cancel_right_isIso]
+  refine ⟨fun hf U ↦ ?_, fun h U ↦ ?_⟩
+  · obtain ⟨hfU, hf⟩ := hf U
+    exact ⟨hfU, by rwa [Scheme.Opens.ι_image_top] at hf⟩
+  · refine ⟨(h U).1, ?_⟩
+    rw [Scheme.Opens.ι_image_top]
+    exact (h U).2
+
+end
+
+section
+
+variable {Q : ∀ {R S : Type u} [CommRing R] [CommRing S], (R →+* S) → Prop}
+
+/-- If `P` is a morphism property affine locally defined by `affineAnd Q`, `P` is stable under
+composition if `Q` is. -/
+lemma HasAffineProperty.affineAnd_isStableUnderComposition {P : MorphismProperty Scheme.{u}}
+    [HasAffineProperty P (affineAnd Q)] (hQ : RingHom.StableUnderComposition Q) :
+    P.IsStableUnderComposition where
+  comp_mem {X Y Z} f g hf hg := by
+    wlog hZ : IsAffine Z
+    · rw [IsLocalAtTarget.iff_of_iSup_eq_top (P := P) _ (iSup_affineOpens_eq_top _)]
+      intro U
+      rw [morphismRestrict_comp]
+      exact this hQ _ _ (IsLocalAtTarget.restrict hf _) (IsLocalAtTarget.restrict hg _) U.2
+    rw [HasAffineProperty.iff_of_isAffine (P := P) (Q := (affineAnd Q))] at hg
+    obtain ⟨hY, hg⟩ := hg
+    rw [HasAffineProperty.iff_of_isAffine (P := P) (Q := (affineAnd Q))] at hf
+    obtain ⟨hX, hf⟩ := hf
+    rw [HasAffineProperty.iff_of_isAffine (P := P) (Q := (affineAnd Q))]
+    exact ⟨hX, hQ _ _ hg hf⟩
+
+/-- If `P` is a morphism property affine locally defined by `affineAnd Q`, `P` is stable under
+base change if `Q` is. -/
+lemma HasAffineProperty.affineAnd_stableUnderBaseChange {P : MorphismProperty Scheme.{u}}
+    [HasAffineProperty P (affineAnd Q)] (hQi : RingHom.RespectsIso Q)
+    (hQb : RingHom.StableUnderBaseChange Q) :
+    P.StableUnderBaseChange :=
+  HasAffineProperty.stableUnderBaseChange
+    (AlgebraicGeometry.affineAnd_stableUnderBaseChange hQi hQb)
+
+/-- If `Q` contains identities and respects isomorphisms (i.e. is satisfied by isomorphisms),
+and `P` is affine locally defined by `affineAnd Q`, then `P` contains identities. -/
+lemma HasAffineProperty.affineAnd_containsIdentities {P : MorphismProperty Scheme.{u}}
+    [HasAffineProperty P (affineAnd Q)] (hQi : RingHom.RespectsIso Q)
+    (hQ : RingHom.ContainsIdentities Q) :
+    P.ContainsIdentities where
+  id_mem X := by
+    rw [eq_targetAffineLocally P, targetAffineLocally_affineAnd_iff hQi]
+    intro U
+    exact ⟨U.2, hQ _⟩
+
+/-- A convenience constructor for `HasAffineProperty P (affineAnd Q)`. The `IsAffineHom` is bundled,
+since this goes well with defining morphism properties via `extends IsAffineHom`. -/
+lemma HasAffineProperty.affineAnd_iff (P : MorphismProperty Scheme.{u})
+    (hQi : RingHom.RespectsIso Q) (hQl : RingHom.LocalizationPreserves Q)
+    (hQs : RingHom.OfLocalizationSpan Q) :
+    HasAffineProperty P (affineAnd Q) ↔
+      ∀ {X Y : Scheme.{u}} (f : X ⟶ Y), P f ↔
+        (IsAffineHom f ∧ ∀ (U : Y.affineOpens), Q (f.app U)) := by
```

Comment:
I don't think we should state things in terms of `forall U : X.affineOpens`.

Reply:
Hmm, I think I can see why, but we do it in many places, e.g. everything around `HasRingHomProperty` is expressed like that (probably because of [AlgebraicGeometry.affineLocally_iff_affineOpens_le](https://leanprover-community.github.io/mathlib4_docs/Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.html#AlgebraicGeometry.affineLocally_iff_affineOpens_le)). I don't care which way we do it, but we should do it consistently.

Reply:
Okay aligned now with the def of `IsAffineHom`.